File: bench/clbg_c_ref/mandelbrot.c
// CLBG mandelbrot — C reference (scalar, no SIMD intrinsics).
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static long count_inside(long grid_n, long max_iter) {
    const double x_min = -2.0, x_max = 1.0;
    const double y_min = -1.5, y_max = 1.5;
    double dx = (x_max - x_min) / (double)grid_n;
    double dy = (y_max - y_min) / (double)grid_n;
    long inside = 0;
    for (long py = 0; py < grid_n; py++) {
        double cy = y_min + (double)py * dy;
        for (long px = 0; px < grid_n; px++) {
            double cx = x_min + (double)px * dx;
            double zx = 0, zy = 0, zx2 = 0, zy2 = 0;
            long iter = 0;
            while (iter < max_iter) {
                if (zx2 + zy2 >= 4.0) { iter = max_iter + 1; }
                else {
                    zy = 2.0 * zx * zy + cy;
                    zx = zx2 - zy2 + cx;
                    zx2 = zx * zx;
                    zy2 = zy * zy;
                    iter++;
                }
            }
            if (iter == max_iter) inside++;
        }
    }
    return inside;
}

int main(int argc, char **argv) {
    long n = (argc > 1) ? atol(argv[1]) : 16000;
    long iter = (argc > 2) ? atol(argv[2]) : 50;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    long r = count_inside(n, iter);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ms = (t1.tv_sec - t0.tv_sec) * 1000.0 + (t1.tv_nsec - t0.tv_nsec) / 1.0e6;
    printf("mandelbrot N=%ld iter=%ld inside=%ld\n", n, iter, r);
    printf("elapsed: %.0fms\n", elapsed_ms);
    return 0;
}

File: bench/clbg_rust_ref/src/mandelbrot.rs
// CLBG mandelbrot — Rust reference (sequential, no SIMD intrinsics).
// Same algorithm as the Soma version. The published CLBG Rust uses
// hand-written SIMD; this version is a fair head-to-head with Soma.

use std::env;
use std::time::Instant;

fn count_inside(grid_n: usize, max_iter: usize) -> usize {
    let x_min: f64 = -2.0;
    let x_max: f64 = 1.0;
    let y_min: f64 = -1.5;
    let y_max: f64 = 1.5;
    let dx = (x_max - x_min) / grid_n as f64;
    let dy = (y_max - y_min) / grid_n as f64;
    let mut inside = 0usize;
    for py in 0..grid_n {
        let cy = y_min + py as f64 * dy;
        for px in 0..grid_n {
            let cx = x_min + px as f64 * dx;
            let mut zx = 0.0f64;
            let mut zy = 0.0f64;
            let mut zx2 = 0.0f64;
            let mut zy2 = 0.0f64;
            let mut iter = 0usize;
            while iter < max_iter {
                if zx2 + zy2 >= 4.0 {
                    iter = max_iter + 1;
                } else {
                    zy = 2.0 * zx * zy + cy;
                    zx = zx2 - zy2 + cx;
                    zx2 = zx * zx;
                    zy2 = zy * zy;
                    iter += 1;
                }
            }
            if iter == max_iter {
                inside += 1;
            }
        }
    }
    inside
}

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(16000);
    let iter: usize = env::args().nth(2).map(|s| s.parse().unwrap()).unwrap_or(50);
    let t0 = Instant::now();
    let r = count_inside(n, iter);
    let elapsed = t0.elapsed();
    println!("mandelbrot N={} iter={} inside={}", n, iter, r);
    println!("elapsed: {:?}", elapsed);
}
